File: LeetCode/0474/golang/lc0474.go
package lc0474

// Runtime complexity: don't know, but lower bound is O(numZeroes * numOnes * len(strs))
// Auxiliary space: O(numZeroes * numOnes)
// Subjective level: hard.
// Solved on: 2025-11-11
func findMaxForm(strs []string, numZeroes int, numOnes int) int {
	// dp[z][o] holds the maximum size of the subset with at most `z` zeroes and `o` ones.
	dp := make([][]int, numZeroes+1)
	for z := range numZeroes + 1 {
		dp[z] = make([]int, numOnes+1)
	}

	for _, str := range strs {
		// Counts the zeroes and the ones manually. Don't want to use strings.Count just for that.
		zeroes := 0
		ones := 0
		for _, ch := range str {
			if ch == '0' {
				zeroes += 1
			} else {
				ones += 1
			}
		}

		for z := numZeroes; z >= zeroes; z-- {
			for o := numOnes; o >= ones; o-- {
				dp[z][o] = max(dp[z][o], dp[z-zeroes][o-ones]+1)
			}
		}
	}

	return dp[numZeroes][numOnes]
}

File: LeetCode/0474/python/lc0474.py
from typing import List

# Runtime complexity: don't know, but lower bound is O(numZeroes*numOnes*len(strs)).
# Auxiliary space complexity: O(numZeroes*numOnes)
# Subjective level: hard.
# Solved on: 2025-11-11
class Solution:
    def findMaxForm(self, strs: List[str], numZeroes: int, numOnes: int) -> int:
        # dp[z][o] holds the maximum size of the subset with at most `z` zeroes and `o` ones.
        dp = [[0] * (numOnes+1) for _ in range(numZeroes+1)]
        for s in strs:
            zeroes = s.count("0")
            ones = len(s) - zeroes

            for z in range(numZeroes, zeroes-1, -1):
                for o in range(numOnes, ones-1, -1):
                    dp[z][o] = max(dp[z][o], dp[z-zeroes][o-ones]+1)
        return dp[numZeroes][numOnes]
